submission/races.go

```diff
@@ -0,0 +1,284 @@
+			sub.cancels[logURL]()
+			sub.cancels[logURL] = nil
+		}
+	}
+}
+
+// returns whether the group has stopped waiting for more SCTs.
+func (sub *safeSubmissionState) groupComplete(groupName string) bool {
+	sub.mu.Lock()
+	defer sub.mu.Unlock()
+	needs, ok := sub.groupNeeds[groupName]
+	if !ok {
+		return true
+	}
+	return needs <= 0
+}
+
+func (sub *safeSubmissionState) collectSCTs() []*AssignedSCT {
+	sub.mu.Lock()
+	defer sub.mu.Unlock()
+	scts := []*AssignedSCT{}
+	for logURL, r := range sub.results {
+		if r != nil && r.sct != nil {
+			scts = append(scts, &AssignedSCT{LogURL: logURL, SCT: r.sct})
+		}
+	}
+	return scts
+}
+
+// postInterval calculates duration for consequent call. For first parallelStart calls duration is 0, while every next one gets additional dur interval.
+func postInterval(idx int, parallelStart int, dur time.Duration) time.Duration {
+	if idx < parallelStart {
+		return time.Duration(0)
+	}
+	return time.Duration(idx+1-parallelStart) * dur
+}
+
+// Submitter is interface wrapping Log-request-response cycle and any processing.
+type Submitter interface {
+	SubmitToLog(ctx context.Context, logURL string, chain []ct.ASN1Cert) (*ct.SignedCertificateTimestamp, error)
+}
+
+// groupRace shuffles logs within the group, submits avoiding duplicate-requests and collects responses.
+func groupRace(ctx context.Context, chain []ct.ASN1Cert, group *ctpolicy.LogGroupInfo, parallelStart int,
+	state *safeSubmissionState, submitter Submitter) groupState {
+	groupURLs := make([]string, 0, len(group.LogURLs))
+	for logURL := range group.LogURLs {
+		groupURLs = append(groupURLs, logURL)
+	}
+
+	type count struct{}
+	counter := make(chan count, len(groupURLs))
+
+	// Randomize the order in which we send requests to the logs in a group
+	// so we maximize the distribution of logs we get SCTs from.
+	for i, urlNum := range rand.Perm(len(groupURLs)) {
+		subCtx, cancel := context.WithCancel(ctx)
+		go func(i int, logURL string) {
+			timeoutchan := time.After(postInterval(i, parallelStart, PostBatchInterval))
+			select {
+			case <-subCtx.Done():
+				counter <- count{}
+				return
+			case <-timeoutchan:
+			}
+			if state.groupComplete(group.Name) {
+				cancel()
+				counter <- count{}
+				return
+			}
+			if firstRequested := state.request(logURL, cancel); !firstRequested {
+				counter <- count{}
+				return
+			}
+			sct, err := submitter.SubmitToLog(subCtx, logURL, chain)
+			// TODO(Mercurrent): verify SCT
+			state.setResult(logURL, sct, err)
+			counter <- count{}
+		}(i, groupURLs[urlNum])
+	}
+	// Wait until either all logs within groups are processed or context is cancelled.
+	for i := 0; i < len(groupURLs); i++ {
+		select {
+		case <-ctx.Done():
+			break
+		case <-counter:
+			if state.groupComplete(group.Name) {
+				break
+			}
+		}
+	}
+	return groupState{Name: group.Name, Success: state.groupComplete(group.Name)}
+}
+
+func parallelNums(groups ctpolicy.LogPolicyData) map[string]int {
```

Comment:
Can you make the name more descriptive? Replacing 'nums' might help.

Reply:
Renamed to -Calls.